consumer/src/main/java/com/flipkart/varadhi/consumer/GroupedMessageSrc.java

```diff
@@ -0,0 +1,160 @@
+package com.flipkart.varadhi.consumer;
+
+import com.flipkart.varadhi.entities.Message;
+import com.flipkart.varadhi.entities.Offset;
+import com.flipkart.varadhi.spi.services.Consumer;
+import com.flipkart.varadhi.spi.services.PolledMessage;
+import com.flipkart.varadhi.spi.services.PolledMessages;
+import lombok.AllArgsConstructor;
+import lombok.RequiredArgsConstructor;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.lang3.mutable.MutableBoolean;
+
+import java.util.*;
+import java.util.concurrent.BlockingDeque;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.concurrent.atomic.AtomicLong;
+
+@RequiredArgsConstructor
+public class GroupedMessageSrc<O extends Offset> implements MessageSrc {
+
+    private final ConcurrentHashMap<String, GroupTracker> allGroupedMessages = new ConcurrentHashMap<>();
+    private final BlockingDeque<String> freeGroups = new LinkedBlockingDeque<>();
+    private final AtomicLong totalInFlightMessages = new AtomicLong(0);
+    private final long maxInFlightMessages = 100; // todo: make configurable
+
+    private final Consumer<O> consumer;
+
+    @Override
+    public CompletableFuture<Integer> nextMessages(MessageTracker[] messages) {
+        if (hasMaxInFlightMessages()) {
+            return replenishAvailableGroups().thenApply(v -> nextMessagesInternal(messages));
+        }
+        return CompletableFuture.supplyAsync(() -> nextMessagesInternal(messages));
+    }
+
+    private int nextMessagesInternal(MessageTracker[] messages) {
+        GroupTracker groupTracker = getGroupTracker();
+        if (null == groupTracker) {
+            return 0;
+        }
+
+        MessageTracker messageTracker = groupTracker.messages.getFirst().nextMessage();
+        messages[0] = new GroupedMessageTracker(messageTracker);
+        return 1;
```

Comment:
we are always returning 1 message?

We should be return many messages all from a unique group.

Reply:
done